File: app/api/execute/route.py
import json
import io
import contextlib
import traceback
from typing import Dict, Any

def create_response(body: Dict[str, Any], status_code: int = 200) -> Dict[str, Any]:
    return {
        "body": json.dumps(body),
        "status": status_code,
        "headers": {
            "Content-Type": "application/json",
            "Access-Control-Allow-Origin": "*",
            "Access-Control-Allow-Methods": "POST, OPTIONS",
            "Access-Control-Allow-Headers": "Content-Type",
        }
    }

def run_code(code: str) -> Dict[str, Any]:
    try:
        output_buffer = io.StringIO()
        error_buffer = io.StringIO()
        
        with contextlib.redirect_stdout(output_buffer):
            with contextlib.redirect_stderr(error_buffer):
                exec(code, {}, {})
                
        return {
            "output": output_buffer.getvalue() or "Code executed successfully"
        }
    except Exception as e:
        error_msg = f"{str(e)}\n{error_buffer.getvalue()}"
        return {"error": error_msg}

def OPTIONS(request):
    return create_response({}, 204)

async def POST(request):
    try:
        body = await request.json()
        if not body or 'code' not in body:
            return create_response({"error": "No code provided"}, 400)

        result = run_code(body['code'])
        return create_response(
            result,
            200 if "output" in result else 400
        )
    except Exception as e:
        return create_response({
            "error": str(e),
            "traceback": traceback.format_exc()
        }, 500)

File: app/api/execute/route.ts
import { NextRequest, NextResponse } from 'next/server';
import { exec } from 'child_process';
import { promisify } from 'util';

const execAsync = promisify(exec);

interface PythonOutput {
  stdout: string;
  stderr: string;
}

export async function POST(req: NextRequest) {
  try {
    const body = await req.json();
    const { code } = body;

    if (!code) {
      return NextResponse.json(
        { error: "No code provided" },
        { status: 400 }
      );
    }

    // Create base64 encoded Python code to pass as argument
    const encodedCode = Buffer.from(code).toString('base64');
    
    try {
      // Execute Python with encoded code
      const { stdout, stderr }: PythonOutput = await execAsync(
        `python -c "import base64; exec(base64.b64decode('${encodedCode}').decode())"`
      );

      if (stderr) {
        return NextResponse.json({ error: stderr }, { status: 400 });
      }

      return NextResponse.json({ output: stdout }, { status: 200 });

    } catch (execError: unknown) {
      const error = execError as Error;
      return NextResponse.json(
        { error: error.message || "Code execution failed" },
        { status: 500 }
      );
    }

  } catch (error: unknown) {
    const err = error as Error;
    return NextResponse.json(
      { error: err.message || "Failed to process request" },
      { status: 500 }
    );
  }
}

export async function OPTIONS() {
  return new NextResponse(null, {
    status: 200,
    headers: {
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Methods': 'POST, OPTIONS',
      'Access-Control-Allow-Headers': 'Content-Type',
    },
  });
}